### ISSUE TITLE
User object does not exist on `getAuth` when `loadUser` is true

### ISSUE THREAD

**Issue by Ehesp** [clerkinc/javascript#927]
- [ ] Review the documentation: https://docs.clerk.dev/
- [ ] Search for existing issues: https://github.com/clerkinc/javascript/issues
- [ ] Go through package changelog files.
- [ ] Provide the Frontend API key from your application dashboard.

## Package + Version

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [x] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`
- [ ] other:

### Version:

```
2.2.11
```

## Description

The [documentation states](https://clerk.dev/docs/quickstarts/get-started-with-remix#enabling-full-server-side-rendering) that when `loadUser` is set to `true` in the root loader, the full user object will be available, however this is not the case.

```ts
export const loader: LoaderFunction = (args) => {
  return rootAuthLoader(
    args,
    { loadUser: true }
  );
};
```

Within another page loader, the following happens:

```ts
export const loader: LoaderFunction = async (args) => {  
  const auth = await getAuth(args);

  if (auth.userId) {
    console.log(auth.user); // null
  }
}
```

When my `userId` is a string (signed in), the `user` object is always `null`.

I can however see that the user is present in the root loader request:

```ts
export const loader: LoaderFunction = (args) => {
  return rootAuthLoader(args, (request) => {
    console.log(request.user); // valid!
    return {};
  }, { loadUser: true });
};
```

This was also reported on Discord a while ago: https://discord.com/channels/856971667393609759/1040268182503571497

**Comment by dimkl** [comment]
@Ehesp thank you for reporting this. I will check if there is some way to make it available in `getAuth()` or if there is any alternative.
The `user` object is available in the SSR components not everywhere in the application (see the following lines found in the docs link you provided):
> This makes the entire User object available to your components during SSR, allowing the whole app to be rendered on the server side.

**Comment by Ehesp** [comment]
Ahh, I assumed it meant every user object. Let me know, it's not a major issue as I can call it again with the sdk, but it does feel like I'm doing an extra request when the data has already been fetched.

**Comment by dimkl** [comment]
@Ehesp Since all loaders are executed in parallel and there is no way (last time i checked) to share data between root loader or other loaders, we have intentionally not populated the `user` in auth. 
The `user|organisation|session` values are populated by making requests to Clerk server and retrieving the values needed. If we had populated them, then we would have to make an api call in each loader using it which would cause a performance degradation to the system without being visible to the developer using it.
To avoid that we have concluded that the developer should fetch the user explicitly from the Clerk server. eg
```javascript
clerkClient.sessions.getSession(sessionId);
clerkClient.users.getUser(userId);
clerkClient.organizations.getOrganization({ organizationId: orgId });
```
We are discussing with the team:
1. to remove `user|organization|session` from the `getAuth()` return value to avoid this confusion
2. check if there is a way in the latest versions of remix to share data
3. add parameters in `getAuth` to explicitly load the resources

For now, i would suggest you avoid using the `auth.user|organisation|session`.

**Comment by Ehesp** [comment]
Thats @dimkl - I mean this totally makes sense and I also don't believe you can share data in loaders like that.

Ok well it's good to know, I can work around it (am doing anyway) but removing them would be better in the long run I think.

Another question I would then have is: What is the use of `loadUser: true` in the root loader:

```ts
export const loader: LoaderFunction = (args) => {
  return rootAuthLoader(args, { loadUser: true });
};
```

What is this actually doing vs it being `false`?

**Comment by dimkl** [comment]
`loadUser: true` It preloads the `user` into the `request` and makes it available to the SSR components. eg
```typescript
export const loader: LoaderFunction = (args) => {
  return rootAuthLoader(
    args,
    ({ request }) => {
      const { user } = request;
      console.log('Loaded user from root loader:', user);
      // the return value will be injected to `request` 
      return { user };
    },
    { loadUser: true },
  );
};
```
To sum i would suggest to use `loadUser: true` in `rootAuthLoader(opts, handler)` if you want to have `user` object in SSR components or if you want to use custom auth logic in the `handler` that depends on the `user` object.
Otherwise i would suggest to omit that param (results in 1 less external request which will increase your application performance when users are signed in).

**Comment by Ehesp** [comment]
Thanks!

When you say "if you want to have user object in SSR components", which components do you mean?

At the moment, in my loaders (other than root loader), I do something like:

```ts
export const loader: LoaderFunction = async (args) => {
  const auth = await getAuth(args);

  if (!auth.userId) {
    return redirect('/login');
  }

  // etc...
}
```

The Clerk components I'm using currently (`UserButton` & `OrganizationSwitcher`) all seem to be client rendered (or at least don't show until JS loads - assume you lazy import them). So I'm a bit confused as to what you mean by SSR components.

**Comment by clerk-cookie** [comment]
This issue is stale because it has been open 30 days with no activity. Remove stale label or comment or this will be closed in 10 days.

**Comment by clerk-cookie** [comment]
This issue was closed because it has been stalled for 10 days with no activity.

**Comment by clerk-cookie** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
